### PR TITLE
CAMS-376 debtor name search

### PR DESCRIPTION
**Pull request by mattstankey**
# Purpose

Implements phonetic and nickname-aware debtor name search functionality using Soundex, Metaphone, and nickname matching algorithms. This allows users to find bankruptcy cases even when debtor names are misspelled, use nicknames (e.g., "Mike" → "Michael"), or have phonetic variations (e.g., "Jon" → "John").   

# Major Changes

  Backend                                                                                                                                                                                                 
  - Phonetic Matching Engine (phonetic-utils.ts):                                                                                                                                                         
    - Soundex and Metaphone phonetic algorithms for fuzzy matching                                                                                                                                        
    - Nickname expansion using name-match library (Mike ↔ Michael, Bob ↔ Robert, etc.)                                                                                                                    
    - Multi-word matching with configurable similarity threshold (0.83)                                                                                                                                   
    - Scoring system: direct match (1.0) → prefix match (0.9) → nickname (0.95) → phonetic (0.83+) → similarity                                                                                           
  - Database Layer (cases.mongo.repository.ts):                                                                                                                                                           
    - MongoDB query using pre-computed phonetic tokens stored on debtor records                                                                                                                           
    - Compound text index on debtor.phoneticTokens and jointDebtor.phoneticTokens                                                                                                                         
    - Post-query filtering with Jaro-Winkler similarity to prevent false positives                                                                                                                        
  - Use Case Layer (case-management.ts):                                                                                                                                                                  
    - Phonetic search filtering moved from repository to use case for better separation                                                                                                                   
    - Feature flag controlled: phonetic-search-enabled                                                                                                                                                    
  - Data Migration (add-phonetic-tokens.ts):                                                                                                                                                              
    - Backfill script to generate phonetic tokens for existing cases                                                                                                                                      
    - Verification script for MongoDB indexes                                                                                                                                                             
                                                                                                                                                                                                          
  Frontend                                                                                                                                                                                                
  - Added "Debtor Name" column to search results table (3rd column)                                                                                                                                       
  - Debtor name search field with 2-character minimum                                                                                                                                                     
  - Visual feedback for phonetic matches                                                                                                                                                                  
  - Feature flag integration              

# Testing/Validation

  Testing                                                                                                                                                                                                 
  - 41 unit tests for phonetic-utils covering:                                                                                                                                                            
    - Nickname matching (Mike/Michael, Bob/Robert, Bill/William, etc.)                                                                                                                                    
    - Phonetic matching (Jon/John, Muhammad/Mohammed, José/Jose)                                                                                                                                          
    - Partial matches ("Mike sm" → "Michael Smith")                                                                                                                                                       
    - False positive prevention (Jon ≠ Jane, Miller ≠ Mike)                                                                                                                                               
    - International name variations and misspellings                                                                                                                                                      
  - BDD specs for end-to-end search scenarios                                                                                                                                                             
  - Mock data with 110+ test cases covering edge cases   

# Notes

  Documentation                                                                                                                                                                                           
  - PHONETIC_SEARCH.md: Implementation guide and algorithm details                                                                                                                                        
  - operations/phonetic-search-indexes.md: MongoDB index setup and maintenance

# Definition of Done:

- [ ] Code refactored for clarity: Developers can understand the work simply by reviewing the code
- [ ] Dependency rule followed: More important code doesn’t directly depend on less important code
- [ ] Development debt eliminated: UX and code aligns to the team’s latest understanding of the domain
- [ ] No regressions: Changes do not cause regression in related or unrelated areas of the application

## Summary by Sourcery

Add phonetic and nickname-aware debtor name search across backend and frontend, including storage of phonetic tokens on cases, hybrid Mongo scoring, and UI support for debtor-name-based searching and display.

New Features:
- Support debtor-name-based case searching using phonetic tokens, bigrams, and nickname expansion with a feature-flag-controlled hybrid search path.
- Store generated phonetic search tokens on debtor and joint debtor records during export-and-load and case sync flows to power fuzzy matching.
- Expose a debtor name search field and optional debtor name column in search and My Cases screens, controlled by feature flags, with validation and analytics metadata.

Bug Fixes:
- Prevent potential log injection or leakage in orders runtime state logging by avoiding logging full state values.

Enhancements:
- Introduce a SCORE stage in the query pipeline and Mongo aggregate renderer to compute weighted phonetic match scores across multiple debtor fields.
- Refine case search tests, repository tests, and case management tests to cover phonetic search behavior, error wrapping, and feature-flag-driven branches.
- Improve LaunchDarkly test infrastructure and feature flag handling to rely on runtime spies instead of hoisted mocks for more reliable flag overrides.
- Add generic trimmed-string validation support and extend search form validation to treat debtor name and whitespace-only input correctly.
- Update case history repository types to use a constrained document type union instead of raw strings for safer usage.

Build:
- Add name-match and natural libraries to backend dependencies to support phonetic and nickname matching.

CI:
- Adjust BDD and React tests (including session timeout warning and feature flag diagnostics) to align with new behavior and React testing best practices.

Documentation:
- Clarify feature flag behavior in tests and configuration comments to document how phonetic search and debtor name column flags are used.

Tests:
- Add extensive unit tests for phonetic token generation, query scoring, export-and-load phonetic enrichment, Mongo aggregate rendering, and case management phonetic search paths.
- Extend frontend tests for search, My Cases, search results rendering, and feature flags to cover debtor name search UX, validation, and conditional columns.
- Introduce reusable phonetic-search case fixtures and dataset-based testing support for end-to-end phonetic search scenarios.

### CODE REVIEW

File: backend/lib/adapters/gateways/mongo/cases.mongo.repository.test.ts
Comment (fmaddenflx) [inline backend/lib/adapters/gateways/mongo/cases.mongo.repository.test.ts]: I wonder if there are some common patterns in this test file that we can break into helper functions or use test.each to condense this test file, as it's getting large.

**Comment by mattstankey** [comment]
@sourcery-ai review

File: backend/lib/adapters/utils/phonetic-helper.test.ts [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+    });
+
+    it('should handle undefined input', () => {
+      const tokens = generatePhoneticTokens(undefined);
+      expect(tokens).toEqual([]);
+    });
+
+    it('should handle names with special characters', () => {
+      const tokens = generatePhoneticTokens("O'Brien");
+      expect(tokens.length).toBeGreaterThan(0);
```
Comment (fmaddenflx) [inline backend/lib/adapters/utils/phonetic-helper.test.ts]: I wonder if we could be more specific with our test here... Do we know what the tokens should look like?

File: user-interface/src/search/SearchResultsHeader.test.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+          <SearchResultsHeader id="test-header" labels={labels} phoneticSearchEnabled={true} />
+        </thead>
+      </table>,
+    );
+
+    expect(screen.getByText('Case Number (Division)')).toBeInTheDocument();
+    expect(screen.getByText('Case Title')).toBeInTheDocument();
+    expect(screen.getByText('Debtor Name')).toBeInTheDocument();
+    expect(screen.getByText('Chapter')).toBeInTheDocument();
+    expect(screen.getByText('Case Filed')).toBeInTheDocument();
```
Comment (fmaddenflx) [inline user-interface/src/search/SearchResultsHeader.test.tsx]: In this and the following test, let's search for the element either by "key" or add a test-id to each of the column headers.  Querying by text is too brittle.

File: backend/lib/adapters/utils/phonetic-helper.test.ts (hunk 1 above)
Comment (diabagatekelly) [reply backend/lib/adapters/utils/phonetic-helper.test.ts]: This test checks integration (that tokens are added during case loading), we test algorithm correctness with specific tokens in phonetic-helper.test.ts.

Comment (mattstankey) [reply backend/lib/adapters/utils/phonetic-helper.test.ts]: This is just testing to see if we can search with an apostrophe at all.  We can expand on it to do a full search, but those are handled in other specs in theory.

File: backend/lib/adapters/utils/phonetic-helper.test.ts [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+            name: 'Jose Garcia',
+            phoneticTokens: generatePhoneticTokens('Jose Garcia'),
+          },
+          documentType: 'SYNCED_CASE',
+        } as SyncedCase,
+      ];
+
+      const filtered = filterCasesByDebtorNameSimilarity(cases, 'John');
+      expect(filtered).toHaveLength(0); // Jose should not match John
+    });
```
[the remark this answers is not in the document]
Comment (fmaddenflx) [reply backend/lib/adapters/utils/phonetic-helper.test.ts]: I think probably using test.each() would be good here if we have several different tests that are doing much the same thing.

**Review by fmaddenflx — APPROVED** [review]

File: backend/lib/adapters/utils/phonetic-helper.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+import { getNameVariations } from 'name-match/src/name-normalizer';
+
+const soundex = new natural.SoundEx();
+const metaphone = new natural.Metaphone();
+
+function normalizeText(text: string): string {
+  return text
+    .trim()
+    .toLowerCase()
+    .replace(/[^a-z\s]/g, '');
```
[the remark this answers is not in the document]
Comment (mattstankey) [reply backend/lib/adapters/utils/phonetic-helper.ts]: Added numbers to the allowed characters.
